src/circuit.rs

```diff
@@ -50,13 +270,401 @@ impl plonk::Circuit<pallas::Base> for Circuit {
+                )?;
+
+                let poseidon_hasher = PoseidonHash::init(
+                    config.poseidon_chip(),
+                    layouter.namespace(|| "Poseidon init"),
+                    ConstantLength::<2>,
+                )?;
+                let poseidon_output = poseidon_hasher.hash(
+                    layouter.namespace(|| "Poseidon hash (nk, rho_old)"),
+                    poseidon_message,
+                )?;
+                let poseidon_output: CellValue<pallas::Base> = poseidon_output.inner.into();
+                poseidon_output
+            };
+
+            // Add hash output to psi using standard PLONK
+            // `scalar` = poseidon_hash(nk, rho_old) + psi_old.
+            //
+            let scalar = {
+                let scalar_val = nk_rho_old
+                    .value()
+                    .zip(psi_old.value())
+                    .map(|(nk_rho_old, psi_old)| nk_rho_old + psi_old);
+                let scalar = self.load_private(
+                    layouter.namespace(|| "poseidon_hash(nk, rho_old) + psi_old"),
+                    config.advices[0],
+                    scalar_val,
+                )?;
+
+                config.plonk_chip().add(
+                    layouter.namespace(|| "poseidon_hash(nk, rho_old) + psi_old"),
+                    nk_rho_old,
+                    psi_old,
+                    scalar,
+                    Some(pallas::Base::one()),
+                    Some(pallas::Base::one()),
+                    Some(pallas::Base::one()),
+                )?;
+
+                scalar
+            };
+
+            // Multiply scalar by NullifierK
+            // `product` = [poseidon_hash(nk, rho_old) + psi_old] NullifierK.
+            //
+            let product = {
+                let nullifier_k = FixedPointBaseField::from_inner(ecc_chip.clone(), NullifierK);
+                nullifier_k.mul(
+                    layouter.namespace(|| "[poseidon_output + psi_old] NullifierK"),
+                    scalar,
+                )?
+            };
+
+            // Add cm_old to multiplied fixed base to get nf_old
+            // cm_old + [poseidon_output + psi_old] NullifierK
+            cm_old
+                .add(layouter.namespace(|| "nf_old"), &product)?
+                .extract_p()
+        };
+
+        // Spend authority
+        // TODO: constrain to equal public input rk
+        let _rk = {
+            // alpha_commitment = [alpha] SpendAuthG
+            let (alpha_commitment, _) = {
+                let spend_auth_g = OrchardFixedBasesFull::SpendAuthG;
+                let spend_auth_g = FixedPoint::from_inner(ecc_chip.clone(), spend_auth_g);
+                spend_auth_g.mul(layouter.namespace(|| "[alpha] SpendAuthG"), self.alpha)?
+            };
+
+            // [alpha] SpendAuthG + ak
+            alpha_commitment.add(layouter.namespace(|| "rk"), &ak)?
+        };
+
+        // Diversified address integrity.
+        let (pk_d_old, _) = {
```

Comment:
The discarded value here is ivk, right? I think @str4d mentioned on the ECC PR that it was a bit strange to just pass through the scalar value from the input to the output. At least add a comment.